packages/date-picker/src/vaadin-date-picker-mixin.js

```diff
@@ -703,27 +702,57 @@ export const DatePickerMixin = (subclass) =>
       }
       if (!dateEquals(this[property], date)) {
         this[property] = date;
-        if (this.value) {
-          this.validate();
-        }
       }
     }
 
-    /** @private */
-    _valueChanged(value, oldValue) {
-      this._handleDateChange('_selectedDate', value, oldValue);
+    /**
+     * Override the value observer from `InputMixin` to implement custom
+     * handling of the `value` property. The date-picker doesn't forward
+     * the value directly to the input like the default implementation of `InputMixin`.
+     * Instead, it parses the value into a date, puts it in `_selectedDate` which
+     * is then displayed in the input with respect to the specified date format.
+     *
+     * @param {string | undefined} newValue
+     * @param {string | undefined} oldValue
+     * @protected
+     * @override
+     */
+    _valueChanged(newValue, oldValue) {
+      const oldSelectedDate = this._selectedDate;
+      this._handleDateChange('_selectedDate', newValue, oldValue);
+      const newSelectedDate = this._selectedDate;
+
+      this._toggleHasValue(!!newValue);
 
-      this._toggleHasValue(!!value);
+      if (oldSelectedDate !== newSelectedDate && newValue && oldValue !== undefined) {
+        this.validate();
+      }
     }
 
-    /** @private */
-    _minChanged(value, oldValue) {
-      this._handleDateChange('_minDate', value, oldValue);
+    /**
+     * @param {string | undefined} newMin
+     * @param {string | undefined} oldMin
+     * @protected
+     */
+    _minChanged(newMin, oldMin) {
+      this._handleDateChange('_minDate', newMin, oldMin);
+
+      if (this.value) {
+        this.validate();
+      }
```

Comment:
Should we declare `min` and `max` as constraints e.g. like in `vaadin-number-field`?

```js
  static get constraints() {
    return [...super.constraints, 'min', 'max'];
  }
```

We could then rely on using `_constraintsChanged` observer logic as in other field components.

Reply:
I guess we could do that but since we still need to call `handleDateChanged`, I wonder if observers will be called in the correct sequence so that the validation will happen only after `_minDate` and `_maxDate` are set.

Let me double-check that.

Reply:
Let's probably do this later since it may affect the conditions when validation happens. For example, validation won't happen on `min` and `max` change anymore even if there is a value.

Reply:
I see, then it's not an option indeed.